### PR TITLE
Use JWT session tokens

### LINKED ISSUE #716: Migrate from `getSession` to `JWT`
## Objective
Use jwt to avoid calling the DB to check the session token is valid

## Tasks

- [x] Configure next-auth to use jwt ([file here](https://github.com/labelflow/labelflow/blob/main/typescript/web/src/pages/api/auth/%5B...nextauth%5D.ts))
- [ ] Investigate and replace the code where needed

### PR DESCRIPTION
**Pull request by bbenoist**
## Work performed

* Enabled the JWT session in next-auth config
* Upgrade next-auth to version 4.0.3

## Results

Sessions should now use JWT to store the authentication information. Which should reduce the amount of DB queries.

## Problems encountered

* As the `user` argument of the next-auth `session` callback is not defined when using JWT, I had to rely on `token.sub` instead of `user.id`. I've made it so we could still disable JWT without having to change the callback.

## Caveats

* ⚠️ Requires `JWT_SECRET` environment variable to be defined on production. (Must be generated with `openssl rand -base64 32`)
* ⚠️ Existing user sessions will be revoked

## Resolved issues

Closes #716 

## Newly raised issues

No

### CODE REVIEW

File: cypress/fixtures/jwt.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { User } from "@prisma/client";
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { encode as jwtEncode } from "next-auth/jwt";
```
Comment (valerian) [inline cypress/fixtures/jwt.ts:3]:
```suggestion
import { encode as jwtEncode, JWT } from "next-auth/jwt";
```

File: cypress/fixtures/jwt.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import { v4 as uuid } from "uuid";
+
+const createJwtToken = (user: User) => {
```
Comment (valerian) [inline cypress/fixtures/jwt.ts:6]:
```suggestion
const createJwtToken = (user: User): JWT => {
```

**Comment by nicOwlas** [comment]
@bbenoist I'm unable to sign-in on the JWT released version. I get an error screen after clicking the send magic link button when trying to login with email. Recording:
Screenshot 2022-01-06 at 09 15 13 1

**Comment by bbenoist** [comment]
@nicOwlas Thank you, that's what i was afraid of. I'll look at it!

**Comment by bbenoist** [comment]
@nicOwlas I've just forgot to add our JWT secret (a unique cryptographic key used to sign the tokens) in vercel. It should be OK once Vercel has finished redeploying the app :)
I've used a different one in production so that we reduce the chances of being leaked.

**Comment by nicOwlas** [comment]
Working great now 👍

**Review by valerian — APPROVED** [review]

**Pull request merged**